statedb: An in-memory database

joamaki opened the pull request:
pkg/statedb provides the Hive with an in-memory database based on immutable radix trees. It supports multiple tables, each with their own indexes. Tables are registered by providing them to the state module as group values. The statedb module provides the central sequencing and snapshotting of the database for doing cross-table atomic transactions.

pkg/statedb/example/main.go provides a fairly full-featured example of the API provided by the statedb module, so I won't repeat that in the description here.

The use-cases for this are any components that have state that is concurrently accessed, potentially via multiple indices. Rather than hand-writing this around Go's builtin maps and mutexes this provides the faciility to implement this in a simpler way, with better safety (no locks for readers!) and performance when the use-case is mostly reading. The table events and "query invalidation" mechanisms provide an alternative to event streams and synchronous APIs to affect changes from one sub-system to another and enable an architecture where work can be done in larger batches and asynchronously without direct coupling between sub-systems.

joamaki committed:
hive: Add cell.Group

This allows grouping cells into a single cell, but without the scoping and
naming that comes with Module().

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
statedb: Add the in-memory database module

pkg/statedb provides the Hive with an in-memory database based on immutable radix
trees. It supports multiple tables, each with their own indexes. Tables are
registered by providing them to the statedb module as group values. The state
module provides the central sequencing and snapshotting of the database for
doing cross-table atomic transactions.

pkg/statedb/example/main.go provides a fairly full-featured example of the API
provided by the statedb module.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

dylandreimerink reviewed (approved):
I am interested to see if this is going to live up the expectation. I have high hopes.

bimmlerd commented on pkg/statedb/cell.go:
```diff
@@ -0,0 +1,34 @@
+// (courtesy of github.com/hashicorp/go-memdb). It adapts the go-memdb library for
+// use with Hive by taking the table schemas as a group values from hive and provides
+// typed API (Table[Obj]) for manipulating tables. As the database is based on an
+// immutable data structure, all objects inserted into the database MUST NOT be mutated!
```
I'm pretty afraid of this causing us "spook action at a distance" bugs. Without really thinking about the feasability - it'd be very nice to have something akin to the deadlock detecting lock library. Mabye a mutation detecting wrapper for objects stored in the DB, enabled through some debug build trickery? Or static analysis that looks at write txns?

joamaki replied:
I'm pretty afraid too, though since we haven't had that many bugs with cache.Store that is similarly considered immutable I'm not super worried. I added a trivial check in the commit to verify that we're not mutating the same object. We can more fancy mutation detection with something like https://github.com/zimmski/go-mutesting or similar libraries. client-go also has similar code to check that cache.Store objects aren't mutated. I'll leave integrating something like that for a later PR.

bimmlerd commented on pkg/statedb/db.go:
```diff
@@ -0,0 +1,111 @@
+	for table := range db.memDB.DBSchema().Tables {
+		iter, err := tx.Get(table, "id")
+		if err != nil {
+			panic(err)
```
is this really panic-worthy?

joamaki replied:
It'll only fail if the table schema is bad (e.g. the mandatory "id" index not defined or faulty). But it doesn't seem like this method should be the one panicing so I rewrote this to return the error and to only marshal one object at a time into a io.Writer rather than all of them in one go.

bimmlerd reviewed (approved):
Excited to see where this goes!

joamaki committed:
statedb: Add benchmarks, change ToJSON into WriteJSON

Add an insert and transaction benchmarks for a rough estimate
on statedb performance. On my laptop I'm getting ~350k insertions
per second (UUID primary key) and ~100k single-insert write transactions
per second.

To support dumping large databases over e.g. REST API, change the ToJSON
into WriteJSON that marshals the objects one by one into the writer rather
than collecting the objects into one large map.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
statedb: Check that modifications are made with a copy

Add a simple form of mutation detection that checks (by pointer comparison)
that on modification to an existing value a DeepCopy()'d object is inserted.

Extend the tests to do modifications. Validated manually that the mutation
detection resulted in panic.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

rolinh reviewed (changes requested):
I think that we can't be adding this dependency: [hashicorp/go-memdb](https://github.com/hashicorp/go-memdb). It is licensed under the Mozilla Public License 2.0 which isn't a license approved by the CNCF (see [approved licenses](https://github.com/cncf/foundation/blob/main/allowed-third-party-license-policy.md#approved-licenses-for-allowlist)). It is also not in the [list of license exceptions](https://github.com/cncf/foundation/tree/main/license-exceptions).

rolinh reviewed (approved):
We discussed the licensing question offline. We will ask the CNCF to add go-memdb to the license exceptions list and expect it to be approved before Cilium v1.14 is released. This particular package is widely used and other popular MPL2.0 libraries from Hashicorp have already been added to the exception list.

pippolo84 commented on pkg/statedb/db_test.go:
```diff
@@ -0,0 +1,254 @@
+			b.Fatalf("Insert error: %s", err)
+		}
+		tx.Commit()
+	}
```
Same as before for `b.StopTimer()`

pippolo84 commented on pkg/statedb/db_test.go:
```diff
@@ -0,0 +1,254 @@
+		}
+	}
+	tx.Commit()
+}
```
It's not much of a difference, but adding a `b.StopTimer()` here we'll remove the accounting of the deferred `hive.Stop()` in the benchmark.

pippolo84 reviewed (approved):
Only a couple of non-blocking nits left inline. LGTM :rocket:
